### PR TITLE
WT-13027 Update assertion to exception in wt_verify.py

### PR DESCRIPTION
**Pull request by zunyiliu**

### CODE REVIEW

File: tools/wt_verify/wt_verify.py [hunk 1]
```diff
@@ -125,7 +126,8 @@ def parse_node(f, line, output, chkpt_info, root_addr, is_root_node):
     node = {}
     line = line[2:-1] # remove new node symbol
     page_type = line.split()[-1]
-    assert page_type == "internal" or page_type == "leaf"
+    if not page_type in ("internal", "leaf"):
```
Comment (etienneptl) [inline tools/wt_verify/wt_verify.py:129]: np, I prefer the array style
```suggestion
    if not page_type in ["internal", "leaf"]:
```

File: tools/wt_verify/wt_verify.py [hunk 2]
```diff
@@ -125,7 +126,8 @@ def parse_node(f, line, output, chkpt_info, root_addr, is_root_node):
...
+        raise Exception(f"page_type expected to be 'internal' or 'leaf' but found '{page_type}'")
```
Comment (etienneptl) [inline tools/wt_verify/wt_verify.py:130]: This suggests you create a variable that contains the two possible values.
```suggestion
        raise Exception(f"page_type expected to be one of the values from {<variable_here>} or 'leaf' but found '{page_type}'")
```

File: tools/wt_verify/wt_verify.py [hunk 3]
```diff
@@ -232,12 +236,14 @@ def parse_dump_pages():
         output = {}
         line = f.readline()
         while line:
-            assert line == SEPARATOR
+            if line != SEPARATOR:
+                raise Exception(f"Expected a separator line but found '{line}'")
```
Comment (etienneptl) [inline tools/wt_verify/wt_verify.py:240]:
```suggestion
                raise Exception(f"Expected {SEPARATOR} but found '{line}'")
```

File: tools/wt_verify/wt_verify.py [hunk 4]
```diff
@@ -586,12 +593,17 @@ def main():
 
     parsed_data = None
 
-    if "dump_pages" in command:
-        parsed_data = parse_dump_pages()
-    elif "dump_blocks" in command:
-        parsed_data = parse_dump_blocks()
-    else:
-        assert False, f"Unexpected command '{command}'"
+    try:
```
Comment (etienneptl) [inline tools/wt_verify/wt_verify.py:596]: After re-reading the code, I think from the `ArgumentParser`, we should never have an issue on the value:
```
    parser.add_argument('-d', '--dump', required=True, choices=['dump_blocks','dump_pages'], help='Option to specify dump_pages configuration.')
```
What are your thoughts on this @korteland ?

**Review by etienneptl — COMMENTED** [review]
Thank you for taking care of this! I have left a few comments.

File: tools/wt_verify/wt_verify.py [hunk 5]
```diff
@@ -177,8 +180,8 @@ def parse_dump_blocks():
         # (i.e > addr: [0: 4096-8192, 4096, 1421166157])
         if is_root:
             x = re.search(r"^> addr: \[\d+: (\d+)-\d+, (\d+), \d+]$", line)
-            assert x, f"Root information expected in '{line}"
-
+            if not x:
+                raise Exception(f"Root information expected in '{line}")
```
Comment (korteland) [inline tools/wt_verify/wt_verify.py:184]: Mismatched single quotes here

File: tools/wt_verify/wt_verify.py [hunk 6]
```diff
@@ -148,7 +150,8 @@ def parse_chkpt_info(f):
     else:
         raise RuntimeError("Could not find checkpoint name")
     line = f.readline()
-    assert line.startswith("Root:")
+    if not line.startswith("Root:"):
+        raise Exception(f"Expected a line starts with 'Root:' but found '{line}'")
```
Comment (korteland) [inline tools/wt_verify/wt_verify.py:154]: Nitpick: "a line" -> "the line"

File: tools/wt_verify/wt_verify.py [hunk 7]
```diff
@@ -586,12 +593,17 @@ def main():
...
+        if "dump_pages" in command:
+            parsed_data = parse_dump_pages()
+        elif "dump_blocks" in command:
+            parsed_data = parse_dump_blocks()
+        else:
+            raise Exception(f"Unexpected command '{command}'")
+    except Exception as e:
+        print(str(e), file=sys.stderr)
+        print(traceback.format_exc(), file=sys.stderr)
+        sys.exit(1)
```
Comment (korteland) [inline tools/wt_verify/wt_verify.py:606]: I don't quite get the logic here - why not remove the `try/except` wrapping and let the exception bubble back up? That way you should get the backtrace for free, and it'll still print the error (I think).

File: tools/wt_verify/wt_verify.py [hunk 8]
```diff
@@ -614,10 +626,15 @@ def main():
...
-        imgs += show_free_block_distribution(args.filename, parsed_data)
-        serve(imgs)
+        try:
+            imgs = show_block_distribution_broken_barh(args.filename, parsed_data)
+            imgs += show_block_distribution_hist(args.filename, parsed_data)
+            imgs += show_free_block_distribution(args.filename, parsed_data)
+            serve(imgs)
+        except Exception as e:
+            print(str(e), file=sys.stderr)
+            print(traceback.format_exc(), file=sys.stderr)
```
Comment (korteland) [inline tools/wt_verify/wt_verify.py:636]: Same thing here - could we just "let it fail" without the `try`/`except` machinery?

File: tools/wt_verify/wt_verify.py (hunk 7 above)
Comment (zunyiliu) [reply tools/wt_verify/wt_verify.py]: Yes the `try/except` is not necessary, I have removed it, thank you!

**Review by etienneptl — APPROVED** [review]
LGTM, nice one!

File: tools/wt_verify/wt_verify.py [hunk 9]
```diff
@@ -588,10 +595,9 @@ def main():
 
     if "dump_pages" in command:
         parsed_data = parse_dump_pages()
-    elif "dump_blocks" in command:
-        parsed_data = parse_dump_blocks()
     else:
-        assert False, f"Unexpected command '{command}'"
+        # If it's "dump_blocks" then do parse_dump_blocks()
+        parsed_data = parse_dump_blocks()
```
Comment (korteland) [inline tools/wt_verify/wt_verify.py:600]: I think the last refactor left this in a bad state - there's no longer any checking that `command` contains something sane, as we'll always fall into the `else` case.

Comment (zunyiliu) [reply tools/wt_verify/wt_verify.py:600]: The parser.add_argument has included a checking `choices=['dump_blocks','dump_pages']` for the command, I think one validation is sufficient as the input is in a single component. What is your opinion? @etienneptl

Comment (etienneptl) [reply tools/wt_verify/wt_verify.py:600]: I agree with you, this should be enough. Should we add a comment saying the argument parser already checks the possible values for future readers?

Comment (etienneptl) [reply tools/wt_verify/wt_verify.py:600]: @korteland, would a comment be enough?

Comment (korteland) [reply tools/wt_verify/wt_verify.py:600]: Sorry - I had a few other things on today and didn't see this thread. I didn't realise the parser already checked this for us, in that case I'm happy with the change as-is :-)

**Review by korteland — APPROVED** [review]
LGTM!

**Pull request merged**